src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+        include(op->value);
+        inside_store = false;
+        include(op->index);
+    }
+
+    bool inside_store;
+    set<string> store_names;
+    Scope<Expr> let_bindings;
+    bool found = false;
+};
+
+class CollectStoreNames : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+    
+    void visit(const Store *op) override {
+        include(op->predicate);
+        include(op->value);
+        include(op->index);
+        store_names.insert(op->name);
+    }
+
+    set<string> store_names;
+};
+
+class RemoveUnnecessaryMutexUse : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    Stmt visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+        // Search for let bindings that access the providers.
+        FindAtomicLetBindings finder(collector.store_names);
+        op->body.accept(&finder);
+        if (finder.found) {
+            // Can't remove mutex lock. Leave the Stmt as is.
+            return IRMutator::visit(op);
+        } else {
+            remove_mutex_lock_names.insert(op->mutex_name);
+            Stmt body = mutate(op->body);
+            return Atomic::make(op->producer_name,
+                                "",
+                                {},
+                                op->tuple_size,
+                                op->dimensions,
+                                std::move(body));
+        }
+    }
+
+    std::set<string> remove_mutex_lock_names;
+};
+
+class FindAtomicMutexUsage : public IRGraphVisitor {
+public:
+    FindAtomicMutexUsage(const std::string &producer_name)
+        : producer_name(producer_name) {}
+
+    using IRGraphVisitor::visit;
+
+    void visit(const Atomic *op) override {
+        for (Expr i : op->mutex_indices) {
+            include(i);
+        }
+        include(op->body);
+        if (op->producer_name == producer_name &&
+                op->mutex_name != "") {
+            if (found) {
+                // Multiple atomics inside the producer,
+                // make sure they have consistent information
+                internal_assert(mutex_name == op->mutex_name &&
+                    tuple_size == op->tuple_size &&
+                    dimensions == op->dimensions) <<
+                    "Inconsistent informatino of atomics inside a Producer node.\n";
+            }
+            found = true;
+            mutex_name = op->mutex_name;
+            tuple_size = op->tuple_size;
+            dimensions = op->dimensions;
+        }
+    }
+
+    const std::string &producer_name;
+    bool found = false;
+    std::string mutex_name;
+    int tuple_size = 0;
+    int dimensions = 0;
+};
+
+class AddAtomicMutex : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    const Target &target;
+
+    AddAtomicMutex(const Target &target) : target(target) {}
+
+    Stmt visit(const ProducerConsumer *op) override {
```

Comment:
What happens with sliding window? Should this instead happen at the Allocate node?

Reply:
Hmm. Directly modifying this code to take place in Allocate doesn't work since there is no allocate node for the output buffer. There indeed is a problem with sliding window that we over-allocate mutex buffer. Let me think.

Reply:
Some. See my comment below.